src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocations.Analyzer.cs

```diff
@@ -0,0 +1,267 @@
+        // We want to offer a diagnostic for this case, so the user explicitly passes the ancestor's ct.
+        private static bool InvocationIgnoresOptionalCancellationToken(IParameterSymbol lastParameter, ImmutableArray<IArgumentOperation> arguments, INamedTypeSymbol cancellationTokenType)
+        {
+            if (lastParameter.Type.Equals(cancellationTokenType) &&
+                lastParameter.IsOptional) // Has a default value being used
+            {
+                // Find out if the ct argument is using the default value
+                return arguments.Any(x =>
+                    x.Parameter.Type.Equals(cancellationTokenType) &&
+                    x.ArgumentKind == ArgumentKind.DefaultValue); // The default value is being used
+            }
+
+            return false;
+        }
+
+        // Checks if the method has a `params CancellationToken[]` argument in the last position and ensure no ct is being passed.
+        private static bool InvocationIsUsingParamsCancellationToken(IParameterSymbol lastParameter, ImmutableArray<IArgumentOperation> arguments, INamedTypeSymbol cancellationTokenType)
+        {
+            if (lastParameter.IsParams &&
+                   lastParameter.Type.Kind == SymbolKind.ArrayType &&
+                   lastParameter.Type is IArrayTypeSymbol arrayTypeSymbol &&
+                   arrayTypeSymbol.ElementType.Equals(cancellationTokenType))
+            {
+                IArgumentOperation? paramsArgument = arguments.FirstOrDefault(a => a.ArgumentKind == ArgumentKind.ParamArray);
+                if (paramsArgument != null)
+                {
+                    if (paramsArgument.Value is IArrayCreationOperation arrayOperation)
+                    {
+                        // Do not offer a diagnostic if the user already passed a ct to the params
+                        return arrayOperation.Initializer.ElementValues.IsEmpty;
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        // Check if there's a method overload with the same parameters as this one, in the same order, plus a ct at the end.
+        private static bool MethodHasCancellationTokenOverload(IMethodSymbol method, ITypeSymbol cancellationTokenType)
+        {
+
+            IMethodSymbol? overload = method.ContainingType.GetMembers(method.Name)
+                                                           .OfType<IMethodSymbol>()
+                                                           .FirstOrDefault(methodToCompare =>
+                                                                HasSameParametersPlusCancellationToken(cancellationTokenType, method, methodToCompare));
+
+            return overload != null;
+
+            // Checks if the parameters of the two passed methods only differ in a ct.
+            static bool HasSameParametersPlusCancellationToken(ITypeSymbol cancellationTokenType, IMethodSymbol originalMethod, IMethodSymbol methodToCompare)
```

Comment:
I think I remember a similar helper somewhere in the codebase.

Reply:
If you know the name and it works for this case, I'll be happy to use it.